spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -318,9 +332,30 @@ protected void doStop() {
 
 		private final AckMode ackMode = getAckMode();
 
+		private final boolean isManualAck = this.ackMode.equals(AckMode.MANUAL);
+
+		private final boolean isManualImmediateAck = this.ackMode.equals(AckMode.MANUAL_IMMEDIATE)
+				|| this.ackMode.equals(AckMode.MANUAL_IMMEDIATE_SYNC);
+
+		private final boolean isAnyManualAck = this.isManualAck || this.isManualImmediateAck;
+
+		private final boolean isRecordAck = this.ackMode.equals(AckMode.RECORD);
+
 		private final boolean syncCommits = KafkaMessageListenerContainer.this.syncCommits;
 
-		private Thread consumerThread;
+		private final long pauseAfter = getPauseAfter();
+
+		private final boolean pauseEnabled = isPauseEnabled();
+
+		private final BlockingQueue<ConsumerRecords<K, V>> recordsToProcess = new LinkedBlockingQueue<>(getQueueDepth());
+
+		private final int queueDepth = getQueueDepth();
+
+		private final BlockingQueue<ConsumerRecord<K, V>> acks = new LinkedBlockingQueue<>();
```

Comment:
Regarding b): A rebalance should restart all the affected partitions at the last committed offset (and I believe that is what the KafkaConsumer does internally) so keeping any data around after the partitions have been revoked would not be correct. You would actually get duplicates as your consumer would seek back to the latest committed offset and then refetch the messages you now have queued up. 

That actually brings up another issue, allowing the set the start offset during a rebalance from an external rebalance listener. This would be for the case where offsets are managed outside kafka and the application needs to set the offsets from another offset store (e.g. mysql). 

Also, a rebalance should be fairly snappy in my opinion. Consider the case of deploying a new version of your software. If you deploy it using a rolling update, you would trigger a lot of rebalances and to allow the system to make progress while updating, the rebalance would need to be fast.